api/org/commcare/api/persistence/SqlHelper.java

```diff
@@ -0,0 +1,224 @@
+                    preparedStatement.setBinaryStream(i + 1, new ByteArrayInputStream((byte[])obj), ((byte[])obj).length);
+                }
+            }
+            int affectedRows = preparedStatement.executeUpdate();
+
+            if (affectedRows == 0) {
+                throw new SQLException("Creating user failed, no rows affected.");
+            }
+
+            try (ResultSet generatedKeys = preparedStatement.getGeneratedKeys()) {
+                if (generatedKeys.next()) {
+                    int id = generatedKeys.getInt(1);
+                    p.setID(id);
+                    return id;
+                } else {
+                    throw new SQLException("Creating user failed, no ID obtained.");
+                }
+            }
+
+        } catch (SQLException e) {
+            System.out.println("e: " + e);
+            e.printStackTrace();
+        }
+        return -1;
+    }
+
+    /**
+     * Update SQLite DB with Persistable p
+     *
+     * @param c          Database Connection
+     * @param storageKey name of table
+     * @param p          peristable to be updated
+     */
+
+    public static void updateId(Connection c, String storageKey, Persistable p) {
+
+        HashMap<String, Object> map = DatabaseHelper.getMetaFieldsAndValues(p);
+
+        String[] fieldnames = map.keySet().toArray(new String[0]);
+        Object[] values = map.values().toArray(new Object[0]);
+
+        Pair<String, String[]> where = org.commcare.modern.database.DatabaseHelper.createWhere(fieldnames, values, p);
+
+        String query = "UPDATE " + storageKey + " SET " + DatabaseHelper.DATA_COL + " = ? WHERE " + where.first + ";";
+
+        PreparedStatement preparedStatement;
+        try {
+            preparedStatement = c.prepareStatement(query);
+
+            byte[] blob = org.commcare.modern.database.TableBuilder.toBlob(p);
+
+            preparedStatement.setBinaryStream(1, new ByteArrayInputStream((byte[])blob), ((byte[])blob).length);
+            /*
+             * We have to do this weird number stuff because 1) our first arg has already been set
+             * (DATA_COL above) and 2) preparedStatement arguments are 1-indexed
+             */
+            for (int i = 2; i < where.second.length + 2; i++) {
+                Object obj = where.second[i - 2];
+                if (obj instanceof String) {
+                    preparedStatement.setString(i, (String)obj);
+                } else if (obj instanceof Blob) {
+                    preparedStatement.setBlob(i, (Blob)obj);
+                } else if (obj instanceof Integer) {
+                    preparedStatement.setInt(i, ((Integer)obj).intValue());
+                } else if (obj instanceof byte[]) {
+                    preparedStatement.setBinaryStream(i, new ByteArrayInputStream((byte[])obj), ((byte[])obj).length);
+                } else if (obj == null) {
+                    preparedStatement.setNull(i, 0);
+                }
```

Comment:
Are you sure any of these `instanceof` branches get used? Test coverage only shows `instanceof String` branch getting executed, The AS linter seems convinced that none of them are satisfied (stupid robots), and I'm really confused by the for-loop indexes. Can you make sure that the strange loop indexing and `instanceof` checks are really necessary? Thanks!